include/seqan/modifier/modifier_iterator.h

```diff
@@ -593,68 +594,56 @@ operator>(ModifiedIterator<THost, TSpec> const & a, ModifiedIterator<THost, TSpe
 
-template <typename THost, typename TSpec, typename TContainer>
-inline bool
-atBegin(ModifiedIterator<THost, TSpec> & me,
-        TContainer const & container)
-{
-    return atBegin(const_cast<ModifiedIterator<THost, TSpec> const &>(me), container);
-}
-
-template <typename THost, typename TSpec, typename TContainer>
+template <typename THost, typename TSpec>
 inline bool
-atBegin(ModifiedIterator<THost, TSpec> const & me,
-        TContainer const & container)
+atBegin(ModifiedIterator<THost, TSpec> const & me)
 {
-    return atBegin(host(me), container);
+    return atBegin(host(me));
 }
 
-template <typename THost, typename TSpec>
+// rooted overload
+template <typename TContainer1, typename TIterator, typename TSpec1, typename TSpec2, typename TContainer2>
 inline bool
-atBegin(ModifiedIterator<THost, TSpec> & me)
+atBegin(ModifiedIterator<Iter<TContainer1, AdaptorIterator<TIterator, TSpec1> >, TSpec2> & me,
+        TContainer2 &)
 {
-    return atBegin(const_cast<ModifiedIterator<THost, TSpec> const &>(me));
+    return atBegin(me);
 }
 
-template <typename THost, typename TSpec>
+template <typename THost, typename TSpec, typename TContainer>
 inline bool
-atBegin(ModifiedIterator<THost, TSpec> const & me)
+atBegin(ModifiedIterator<THost, TSpec> & me,
+        TContainer & container)
 {
-    return atBegin(host(me));
+    return me == begin(container, Standard());
 }
 
 // --------------------------------------------------------------------------
 // Function atEnd()
 // --------------------------------------------------------------------------
 
-template <typename THost, typename TSpec, typename TContainer>
-inline bool
-atEnd(ModifiedIterator<THost, TSpec> & me,
-      TContainer const & container)
-{
-    return atEnd(const_cast<ModifiedIterator<THost, TSpec> const &>(me), container);
-}
-
-template <typename THost, typename TSpec, typename TContainer>
+template <typename THost, typename TSpec>
 inline bool
-atEnd(ModifiedIterator<THost, TSpec> const & me,
-      TContainer const & container)
+atEnd(ModifiedIterator<THost, TSpec> const & me)
 {
-    return atEnd(host(me), container);
+    return atEnd(host(me));
 }
 
-template <typename THost, typename TSpec>
+// rooted overload
+template <typename TContainer1, typename TIterator, typename TSpec1, typename TSpec2, typename TContainer2>
 inline bool
-atEnd(ModifiedIterator<THost, TSpec> & me)
+atEnd(ModifiedIterator<Iter<TContainer1, AdaptorIterator<TIterator, TSpec1> >, TSpec2> & me,
+      TContainer2 &)
 {
-    return atEnd(const_cast<ModifiedIterator<THost, TSpec> const &>(me));
+    return atEnd(me);
 }
 
-template <typename THost, typename TSpec>
+template <typename THost, typename TSpec, typename TContainer>
 inline bool
-atEnd(ModifiedIterator<THost, TSpec> const & me)
+atEnd(ModifiedIterator<THost, TSpec> & me,
+      TContainer & container)
 {
-    return atEnd(host(me));
+    return me == end(container, Standard());
```

Comment:
See above. The extra rioted overload is needed, because we cannot subtract rooted and standard iterators.